### DISCUSSION TITLE
Buderus Hydraulic Box VC0-VW1 vc0 switch doesn't work

### DISCUSSION THREAD

**Discussion by jimmyhenne** [emsesp/EMS-ESP32#1898] [Q&A]
Hi,

I have a Buderus WLW196i heatpump and use a BBQKees Gateway V2 to integrate it into my Home Assistant system. The heating system also has a threeway valve called 'Hydraulic Box VC0-VW1' with the two valves vc0 and vw1. It's this one:
![vc0vw1](https://github.com/user-attachments/assets/3d6151b5-0e6a-42c4-b9e5-0c0b32960b28)
In Home Assistant is a switch called 'switch.boiler_vc0valve' but it unfortunately doesn't work. It's always off. When the valve changes its position to heat up water, the switch remains at off. And when I manually click the switch, the valve does nothing.

1. Is there something not configured correctly in my system or is this an ems-esp bug?
2. I can't find an entity that contains vw1 to control the vw1 valve. How is it called?

I checked the log when using the vc0 switch in HA. There are no errors or warning, I can't really figure it out but for me there is just one info.

Log:
> ...
> 001+22:30:38.466   TRACE 365:    [emsesp]     boiler(0x08) -B-> All(0x00), ?(0x04AA), data: 00 00 (offset 24)
> 001+22:30:38.830   TRACE 366:    [emsesp]     boiler(0x08) -B-> All(0x00), ?(0x04AB), data: 00 00 07 D0 23 28 00 00 00 00 00 00 00 00
> 001+22:30:39.006   TRACE 367:    [emsesp]     boiler(0x08) -B-> All(0x00), ?(0x04AC), data: 00 00 03
> 001+22:30:43.504   TRACE 368:    [emsesp]     boiler(0x08) -B-> All(0x00), UBAMonitorFastPlus(0xE4), data: 10 20 30 48 00 CB 00 00 E1 00 00 00 00 02 B2 00 00 01 1B 00 00 00 00 01 48 00 E1
**> 001+22:30:43.630    INFO 369:    [command]    Calling command 'boiler/vc0valve' (VC0 Ventil) with value an**
> 001+22:30:43.693   TRACE 370:    [emsesp]     boiler(0x08) -B-> All(0x00), UBAMonitorSlowPlus2(0xE3), data: 04 00 00 00 00 00 00 00 00 00 00 00 E1 00 64 55
> 001+22:30:43.888   TRACE 371:    [emsesp]     boiler(0x08) -B-> All(0x00), UBAMonitorWWPlus(0xE9), data: 00 01 A7 01 72 00 00 00 00 41 3C 00 00 00 00 00 00 00 00 00 00 00 00 00
> 001+22:30:44.103   TRACE 372:    [emsesp]     Me(0x0B) -W-> boiler(0x08), HPSettings3(0x049D), data: 01 (offset 3)
> 001+22:30:44.290   TRACE 373:    [emsesp]     boiler(0x08) -B-> All(0x00), ?(0x049F), data: 00 00 00 00 00 00 00 00 00 00 00 00 00 00
> 001+22:30:44.469   TRACE 374:    [emsesp]     boiler(0x08) -B-> All(0x00), ?(0x04A0), data: 00 00 01 6D
> 001+22:30:44.667   TRACE 375:    [emsesp]     boiler(0x08) -B-> All(0x00), ?(0x04A3), data: 00 01 00 00 00 00 00 00 00 00 00 04 00 00 00 00 00 00 00 00 08 00 00 00
> 001+22:30:44.916   TRACE 376:    [emsesp]     Me(0x0B) -R-> boiler(0x08), HPSettings3(0x049D), length: 0x19
> 001+22:30:44.940   TRACE 377:    [emsesp]     boiler(0x08) -W-> Me(0x0B), HPSettings3(0x049D), data: 00 00 00 01 00 00 00 00 00 00 00 00
> 001+22:30:52.208   TRACE 378:    [emsesp]     thermostat(0x10) -R-> boiler(0x08), ?(0xF7), length: 0x01 (offset 3)
> ...

Systeminfo:

> {
>   "System Info": {
>     "version": "3.6.5",
>     "uptime": "001+22:25:13.157",
>     "uptime (seconds)": 167113,
>     "platform": "ESP32",
>     "arduino": "Tasmota Arduino v2.0.14",
>     "sdk": "4.4.6.240105",
>     "free mem": 188,
>     "max alloc": 107,
>     "used app": 1591,
>     "free app": 4105,
>     "partition": "app0",
>     "reset reason": "Power on reset / APP CPU reset by PRO CPU"
>   },
>   "Network Info": {
>     "network": "Ethernet",
>     "hostname": "ems-esp",
>     "TxPower setting": 78,
>     "static ip config": false,
>     "enable IPv6": true,
>     "low bandwidth": false,
>     "disable sleep": false,
>     "enable MDNS": true,
>     "enable CORS": false,
>     "AP provision mode": "disconnected",
>     "AP security": "wpa2",
>     "AP ssid": "ems-esp"
>   },
>   "NTP Info": {
>     "NTP status": "disconnected",
>     "enabled": false,
>     "server": "time.google.com",
>     "tz label": "Europe/Amsterdam"
>   },
>   "OTA Info": {
>     "enabled": false,
>     "port": 8266
>   },
>   "MQTT Info": {
>     "MQTT status": "connected",
>     "MQTT publishes": 69564,
>     "MQTT queued": 0,
>     "MQTT publish fails": 0,
>     "MQTT connects": 4,
>     "enabled": true,
>     "client id": "ems-esp",
>     "keep alive": 60,
>     "clean session": true,
>     "entity format": 1,
>     "base": "ems-esp",
>     "discovery prefix": "homeassistant",
>     "discovery type": 0,
>     "nested format": 1,
>     "ha enabled": true,
>     "mqtt qos": 0,
>     "mqtt retain": false,
>     "publish time heartbeat": 60,
>     "publish time boiler": 10,
>     "publish time thermostat": 10,
>     "publish time solar": 10,
>     "publish time mixer": 10,
>     "publish time other": 10,
>     "publish time sensor": 10,
>     "publish single": false,
>     "publish2command": false,
>     "send response": false
>   },
>   "Syslog Info": {
>     "enabled": false
>   },
>   "Sensor Info": {
>     "temperature sensors": 0,
>     "temperature sensor reads": 0,
>     "temperature sensor fails": 0,
>     "analog sensors": 0,
>     "analog sensor reads": 0,
>     "analog sensor fails": 0
>   },
>   "API Info": {
>     "API calls": 1,
>     "API fails": 0
>   },
>   "Bus Info": {
>     "bus status": "connected",
>     "bus protocol": "Buderus",
>     "bus telegrams received (rx)": 544127,
>     "bus reads (tx)": 133791,
>     "bus writes (tx)": 11,
>     "bus incomplete telegrams": 0,
>     "bus reads failed": 0,
>     "bus writes failed": 0,
>     "bus rx line quality": 100,
>     "bus tx line quality": 100
>   },
>   "Settings": {
>     "board profile": "E32V2",
>     "locale": "de",
>     "tx mode": 1,
>     "ems bus id": 11,
>     "shower timer": false,
>     "shower alert": false,
>     "hide led": false,
>     "notoken api": false,
>     "readonly mode": false,
>     "fahrenheit": false,
>     "dallas parasite": false,
>     "bool format": 1,
>     "bool dashboard": 1,
>     "enum format": 1,
>     "analog enabled": true,
>     "telnet enabled": true,
>     "max web log buffer": 50,
>     "web log buffer": 50
>   },
>   "Devices": [
>     {
>       "type": "boiler",
>       "name": "Enviline/Compress 6000AW/Hybrid 3000-7000iAW/SupraEco/Geo 5xx/WLW196i/WSW196i",
>       "device id": "0x08",
>       "product id": 172,
>       "version": "01.21",
>       "entities": 148,
>       "handlers received": "0xBF 0xC2 0xD1 0xE3 0xE4 0xE5 0xE9 0x0494 0x0495 0x048F",
>       "handlers fetched": "0x14 0xE6 0xEA 0x048D 0x048A 0x04A2 0x0485 0x0486 0x0492 0x0488 0x0484 0x048B 0x0491 0x0499 0x049C 0x049D 0x02CC",
>       "handlers pending": "0x10 0x11 0x15 0x1C 0x18 0x19 0x1A 0x35 0x16 0x33 0x34 0x26 0x2A 0x28 0x04AE 0x04AF",
>       "handlers ignored": "0x049F 0x04A0 0x04A3 0x04AA 0xF7 0xE7 0x02E0 0x02EA 0xD2 0xE2 0x061E 0x0240 0x029B 0x04AB 0x04AC 0x0497 0xF9 0x2E 0x3B 0x17 0x36 0xF6"
>     },
>     {
>       "type": "thermostat",
>       "name": "RC300/RC310/Moduline 3000/1010H/CW400/Sense II/HPC410",
>       "device id": "0x10",
>       "product id": 158,
>       "version": "73.03",
>       "entities": 100,
>       "handlers received": "0x06 0x02BB 0x02BC 0x02BD 0x02BE 0x02BF 0x02C0 0x031D 0x0267",
>       "handlers fetched": "0x02A5 0x02B9 0x02AF 0x029B 0x0471 0x02A6 0x02BA 0x02B0 0x029C 0x0472 0x02CC 0x0467 0x0291 0x02CE 0x0468 0x0292 0x0293 0x0294 0x02F5 0x023A 0x0240 0xBB 0x023E",
>       "handlers pending": "0xA3 0xA2 0x12 0x13 0x02A7 0x02B1 0x029D 0x0473 0x02A8 0x02B2 0x029E 0x0474 0x02A9 0x02B3 0x029F 0x0475 0x02AA 0x02B4 0x02A0 0x0476 0x02AB 0x02B5 0x02A1 0x0477 0x02AC 0x02B6 0x02A2 0x0478 0x02D0 0x0469 0x02D2 0x046A 0x031B 0x031E",
>       "handlers ignored": "0x0239 0x0508 0x0509 0x050A 0x0507 0xBF 0xF7"
>     },
>     {
>       "type": "mixer",
>       "name": "MM100",
>       "device id": "0x21",
>       "product id": 160,
>       "version": "24.05",
>       "entities": 7,
>       "handlers received": "0x02D8",
>       "handlers fetched": "0x02CE",
>       "handlers ignored": "0x02E2 0x02EC 0xE6 0x0255 0x0292 0xBF 0xF9"
>     }
>   ]
> }

**Comment by MichaelDvP** [comment]
Please check https://emsesp.github.io/docs/Troubleshooting/#im-missing-certain-data-from-an-ems-device
The valve was added in #892, there you can also see how to get the right telegram. 
Setting it on the boiler while logging give you also the info that the lable VC0 is really connected to VC0 on the  boiler clamps.

**Comment by jimmyhenne** [comment]
Thank you for supporting!

I used the function test of the heat pump to test and move the valves and logged the events. First of all VC0 and VW1 are properly connected. When I change VC0 the VC0 valve moves and when I change VW1 (it's called threeway valve in the settings) the vw1 valve moves.

The logs are difficult for me to understand but I think I found the interesting events:
**Three way:**
```
000+02:37:05.930 TRACE 331: [emsesp] thermostat(0x10) -W-> boiler(0x08), HPSettings3(0x049D), data: 01 (offset 6)
000+02:37:07.501 TRACE 341: [emsesp] boiler(0x08) -B-> All(0x00), HPSettings3(0x049D), data: 00 00 00 00 00 00 01 00 00 00 00 00

000+02:37:26.280 TRACE 438: [emsesp] thermostat(0x10) -W-> boiler(0x08), HPSettings3(0x049D), data: 00 (offset 6)
000+02:37:27.215 TRACE 442: [emsesp] boiler(0x08) -B-> All(0x00), HPSettings3(0x049D), data: 00 00 00 00 00 00 00 00 00 00 00 00
```
**VC0:**
```
000+02:37:48.792 TRACE 482: [emsesp] thermostat(0x10) -W-> boiler(0x08), HPSettings3(0x049D), data: 01 (offset 3)
000+02:37:49.601 TRACE 484: [emsesp] boiler(0x08) -B-> All(0x00), HPSettings3(0x049D), data: 00 00 00 01 00 00 00 00 00 00 00 00

000+02:38:08.630 TRACE 558: [emsesp] thermostat(0x10) -W-> boiler(0x08), HPSettings3(0x049D), data: 00 (offset 3)
000+02:38:09.513 TRACE 568: [emsesp] boiler(0x08) -B-> All(0x00), HPSettings3(0x049D), data: 00 00 00 00 00 00 00 00 00 00 00 00
```

When I compare these lines to the ones in #892 I see 2 differences: the words _thermostat_ and _boiler_ are written lower case and my lines contain _HPSettings3_ instead of _?_
Does this make the difference that it doesn't work for me?

[log-emsesp-vw1-vc0.txt](https://github.com/user-attachments/files/16395997/log-emsesp-vw1-vc0.txt)

**Comment by MichaelDvP** [comment]
>but I think I found the interesting events:

Yes, good catch. But this are exactly the bytes we use for the valve detection. After thermostat sends to boiler offset 6 to 01 (and boiler shows full telegram with 01 in offset 6) also emsesp should show 3-way-valve set.

If you set the valve via emsesp it's possibe that the thermostat resets it and you do not see an effect. 
Your first log shows for VC0: command called, set 01 to offset 3, boiler replys with 01 in offset 3
```
> 001+22:30:43.630 INFO 369: [command] Calling command 'boiler/vc0valve' (VC0 Ventil) with value an
001+22:30:44.103 TRACE 372: [emsesp] Me(0x0B) -W-> boiler(0x08), HPSettings3(0x049D), data: 01 (offset 3)
001+22:30:44.940 TRACE 377: [emsesp] boiler(0x08) -W-> Me(0x0B), HPSettings3(0x049D), data: 00 00 00 01 00 00 00 00 00 00 00 
```
Log a bit longer an check if the thermostat or another device takescontrol and resets this value.

**Comment by jimmyhenne** [comment]
Now I have noticed that it only works when the function test of the heater is activated. Then and just then I can toggle the valves manually. I'm afraid the heat pump doesn't allow to change the valves by hand, except for the manual unhooking of the motor and switching directly at the valve.

**Reply by MichaelDvP** [reply]
But there seems to be a state to read. The log shows that 48D offset 0 changes with the setting. Shows 80 for VC0 switched and 40 for VW1. Try to add a custom entity with device 0x08, type 0x48D, offset 0, UINT8. and see if it changes when the valve switch.

**Reply by jimmyhenne** [reply]
This is the custom entity:
![grafik](https://github.com/user-attachments/assets/4a2d7074-105c-4d1d-a900-eb0419bc37a6)

Without doing anything the value was 64, when I initiated water heating the valves change (both) and the value also changed to 129, when it changed to the next mode and VW1 moves, the value changed to 193. After heating resp. cancelling heating, it goes back to 64.
I can't interpret the values 😄

**Reply by MichaelDvP** [reply]
This byte is a bit-flield, bit7 (128) is VC0 on, bit6 (64) is VW1 and something switches with bit0 (1). 
I'll change the valves vc0valve and 3wayvalve to read form 0x48D telegram and write to 0x49D. 
Do you know what the bit0(1) is? a pump? 
I don't have a heatpump and don't know your hydraulic scheme, so also have no idea what these valves do in your system.

**Reply by jimmyhenne** [reply]
Sorry, I don't know what bit0 is. But it may well be the main heating pump, because it also comes on when warm water is being produced.

Edit: I would like to be able to control the hydraulic box and the valve settings by myself. I logged when warm water heating starts and the hydraulic box moves the valves. But the log file overwhelms me. Because two valves moved and there are two similar entries could it be HPSet(0x0467) and HPSet(0x0468) in traces 156-163?

[log240728_ww_start.txt](https://github.com/user-attachments/files/16404935/log240728_ww_start.txt)

**Reply by MichaelDvP** [reply]
You can try build: https://github.com/MichaelDvP/EMS-ESP32/releases
This should show the valve states. 
Setting the valves only works in testmode. Normally there is a function inside heating that controlles the valves and you can only trigger a function. In this case it's form the log:
Trigger dhw charge on the thermostat: `Me(0x0B) -W-> thermostat(0x10), RC300WWmode(0x02F5), data: FF (offset 11)`
Thermostat confirms: `thermostat(0x10) -B-> All(0x00), RC300WWmode(0x02F5), data: FF (offset 11)`
Thermostat sets DHW extra: `thermostat(0x10) -B-> All(0x00), RC300WWmode2(0x031D), data: 01 00 02 02`
Thermostat command the boiler to charge: `thermostat(0x10) -W-> boiler(0x08), ?(0xE7), data: 01 00 01`
Boiler do all that is needed for charging: start the pump, open valves, etc.
